include/proteus/CoreLLVMDevice.hpp

```diff
@@ -234,6 +237,129 @@ inline void specializeIR(Module &M, StringRef FnName, StringRef Suffix,
   runCleanupPassPipeline(M);
 }
 
+inline std::unique_ptr<Module> cloneKernelFromModule(Module& M, const std::string& name) {
+      auto KernelModule = std::make_unique<Module>("JitModule", M.getContext());
+      KernelModule->setSourceFileName(M.getSourceFileName());
+      KernelModule->setDataLayout(M.getDataLayout());
+      KernelModule->setTargetTriple(M.getTargetTriple());
+      KernelModule->setModuleInlineAsm(M.getModuleInlineAsm());
+      KernelModule->IsNewDbgInfoFormat = M.IsNewDbgInfoFormat;
+
+      auto KernelFunction =
+          M.getFunction(name);
+
+      SmallPtrSet<Function *, 8> ReachableFunctions;
+      SmallPtrSet<GlobalVariable *, 16> ReachableGlobals;
+      SmallPtrSet<Function *, 8> ReachableDeclarations;
+      SmallVector<Function *, 8> ToVisit;
+      ReachableFunctions.insert(KernelFunction);
+      ToVisit.push_back(KernelFunction);
+      CallGraphWrapperPass CG;
+      CG.runOnModule(M);
+      while (!ToVisit.empty()) {
+        Function *VisitF = ToVisit.pop_back_val();
+        CallGraphNode *CGNode = CG[VisitF];
+
+        for (const auto &Callee : *CGNode) {
+          Function *CalleeF = Callee.second->getFunction();
+
+          if (!CalleeF)
+            continue;
+
+          if (CalleeF->isDeclaration()) {
+            ReachableDeclarations.insert(CalleeF);
+            continue;
+          }
+
+          if (ReachableFunctions.contains(CalleeF))
+            continue;
+
+          ReachableFunctions.insert(CalleeF);
+          ToVisit.push_back(CalleeF);
+        }
+      }
+
+      auto process_instruction = [&](GlobalVariable &GV, const Instruction *I) {
+        const Function *ParentF = I->getParent()->getParent();
+        if (ReachableFunctions.contains(ParentF))
+          ReachableGlobals.insert(&GV);
+      };
+
+      for (auto &GV : M.globals()) {
+        for (const User *Usr : GV.users()) {
+          const Instruction *I = dyn_cast<Instruction>(Usr);
+
+          if (I) {
+            process_instruction(GV, I);
+          } else {
+            for (const User *NextUser : Usr->users()) {
+              I = dyn_cast<Instruction>(NextUser);
+              if (I)
+                process_instruction(GV, I);
```

Comment:
Where does this need for second-level user iteration come from? Is it because of casting? We should generalize with `Worklist` but I'm OK to do this in a refinement iteration, but please add comment